source/language/types.rst

```diff
@@ -373,3 +375,152 @@ variables whose values may only be known at run time.
++==============+============+============+============+=============+============+============+============+============+
+| **bool**     | :center:`-`| :gbg:`Yes` | :gbg:`Yes` | :gbg:`Yes`  | :rbg:`No`  | :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **int**      | :gbg:`Yes` | :center:`-`| :gbg:`Yes` | :gbg:`Yes`  | :rbg:`No`  | :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **uint**     | :gbg:`Yes` | :gbg:`Yes` | :center:`-`| :gbg:`Yes`  | :rbg:`No`  | :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **float**    | :gbg:`Yes` | :gbg:`Yes` | :gbg:`Yes` | :center:`-` | :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **angle**    | :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  | :gbg:`Yes`  | :center:`-`| :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **bit**      | :gbg:`Yes` | :gbg:`Yes` | :gbg:`Yes` | :rbg:`No`   | :gbg:`Yes` | :center:`-`| :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **duration** | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No*`  | :rbg:`No`  | :rbg:`No`  | :center:`-`| :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **qubit**    | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No`   | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :center:`-`|
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+
+\*Note: ``duration`` values can be converted to ``float`` using the division operator. See :ref:`divideDuration`
+
+Casting from bool
+~~~~~~~~~~~~~~~~~
+
+``bool`` values cast from ``false`` to ``0.0`` and from ``true`` to ``1.0`` or 
+an equivalent representation. ``bool`` values can only be cast to ``bit[1]`` 
+(a single bit), so explicit slicing syntax must be given if the target ``bit``
+has more than 1 bit of precision.
+
+Casting from int/uint
+~~~~~~~~~~~~~~~~~~~~~
+
+``int[n]`` and ``uint[n]`` values cast to the standard types mimicking ``C``
+behavior. Casting to ``bit[m]`` is only allowed when ``m==n``. If the target 
+``bit`` has more or less precision, then explicit slicing syntax must be given.
+
+Casting from float
+~~~~~~~~~~~~~~~~~~
+
+``float[n]`` values cast to the standard types mimicking ``C`` behavior. 
```

Comment:
It might be better just to be explicit and say "discards the fractional part" (for integral types).  Did we decide on what happens if the integral part can't be represented by the new type?

Also:

```suggestion
``float[n]`` values cast to the standard types mimicking C behavior. 
```

Reply:
Added in https://github.com/Qiskit/openqasm/pull/268/commits/93e759e1cb14a14c5de6fdae55ffd33ae963ecbf.

In line with C behavior, I believe it's the same behavior as casting from float to a large enough int, then casting from the large int to the target type.

Reply:
That shakes out as being modulo `UINTX_MAX + 1`, for unsigned, right?  That's sensible.  I think conversion to signed integer types in C is implementation-defined, though, if it doesn't fit - do we want to be prescriptive in that case?

Reply:
You're right, it is implementation defined. I think it would make sense to apply the same approach to the OpenQASM 3 spec. I will add notes regarding this in the appropriate places.

Reply:
Added in https://github.com/Qiskit/openqasm/pull/268/commits/fa1954277231cad03133dc4a1fc0f23b3ffe45c3